### ISSUE TITLE
Half precision training very slow and returning nan loss 

### ISSUE THREAD

**Issue by acmilannesta** [tensorflow/tensorflow#37130]
<em>Please make sure that this is an issue related to performance of TensorFlow.
As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:performance_template</em>

**System information** 

- OS Platform and Distribution : Win10
- TensorFlow version (use command below): 2.1.0
- Python version: 3.6.8
- CUDA/cuDNN version:  10.1 / 7.6.4
- GPU model and memory: RTX2080ti (11GB)

**Describe the current behavior**

I used the "mixed_float16" policy to train the efficientnet model (https://github.com/qubvel/efficientnet), but the training become almost 10 times slower and return nan even if I set a large epsilon.

https://www.tensorflow.org/api_docs/python/tf/keras/mixed_precision/experimental/Policy?hl=en

here are some of the codes to setup mixed float training:
```
from tensorflow.keras.mixed_precision import experimental as mixed_precision
policy = mixed_precision.Policy('mixed_float16')
mixed_precision.set_policy(policy)
K.set_epsilon(1e-3)
```
P.S.: When I trained the densenet121 from (tf.keras.applications) using mixed float16, it runs pretty well.

**Comment by reedwm** [comment]
Can you provide a full example, which creates and runs the efficient net model? I'm not sure how to run efficient net

**Comment by acmilannesta** [comment]
Sure, I made a colab file.
https://colab.research.google.com/drive/1tQbhe_gGhIFoMuHu-HqazHzlP0Z83kwj

**Comment by reedwm** [comment]
On the line `out = keras.layers.BatchNormalization()(x)`, I'm getting the error "name 'x' is not defined". There are other undefined values as well, such as `BATCH_SIZE`.

**Comment by acmilannesta** [comment]
Sorry, I just changed the code.

There are two data files needed "train.csv" and "train_origin2.zip". 
All could be downloaded from https://drive.google.com/drive/folders/1abgTg2uuWBmo2TTvipqiXVhxkBuLmNwb?usp=sharing

**Comment by reedwm** [comment]
Thanks for filing the issue! I could reproduce the slowness on TF 2.1 but not TF 2.2rc0, so I'm assuming the issue has been fixed. I also could not reproduce the NaN issue on 2.2rc0 after waiting for 1 epoch, but did not try on 2.1 since I got impatient waiting. Let me know if you still see the NaN issue.

You can get the fix either by installing 2.2rc0 with `pip install tensorflow==2.2rc0` or by waiting for 2.2 to come out.

**Comment by anuragrs** [comment]
@reedwm Could you post the root cause for slowness on TF 2.1 please? Thanks.

**Comment by reedwm** [comment]
> @reedwm Could you post the root cause for slowness on TF 2.1 please? Thanks.

I'm not sure what the root cause was. Maybe c476ffbbd8e35ac8683cc339802bfd9cdcdaa7a9?

**Comment by Arktius** [comment]
> Check out the loss scaler [TF](https://www.tensorflow.org/api_docs/python/tf/keras/mixed_precision/experimental/LossScaleOptimizer) and the [mixed precision guide](https://www.tensorflow.org/guide/keras/mixed_precision#loss_scaling) .

https://github.com/tensorflow/tensorflow/issues/38495#issuecomment-618682103

Why did you set Keras' epsilon?

**Comment by Hazarapet** [comment]
Same thing. 
Mixed precision with `mixed_float16` is super slow. **10 times slower than without it**
I've used the same batch size, the same model and started to count the time for 

**1. Feedforward process ~10x worst
2. Loss computing process ~6x worst
3. Gradient computing process ~6x works**

All points get worst

**Comment by Arktius** [comment]
> Same thing.
> Mixed precision with `mixed_float16` is super slow. **10 times slower than without it**
> I've used the same batch size, the same model and started to count the time for
> 
> **1. Feedforward process ~10x worst 2. Loss computing process ~6x worst 3. Gradient computing process ~6x works**
> 
> All points get worst

Increase the network's size. I've spent hours to figure out that MP is faster for big-huge networks. What model architecture are you using?

**Comment by Hazarapet** [comment]
@Arktius 
I'm using our designed network, which consists of regular layers like Conv2D, BatchNorm, Activation, etc..
It is an autoencoder, which consists of a pre-trained encoder (vgg, resnet, etc..) and 1-3 decoders.
**I've figured out the problem is related to bilinear upsampling (inside Upsample2D or tf.image.resize) and it breaks the mixed-precision performance.**
**It is too bad because the same network implemented with another framework is ~7-8x faster**

**Comment by Arktius** [comment]
> @Arktius
> I'm using our designed network, which consists of regular layers like Conv2D, BatchNorm, Activation, etc..
> It is an autoencoder, which consists of a pre-trained encoder (vgg, resnet, etc..) and 1-3 decoders.
Do you train the whole network or just the last couple of layers? (Transfer Learning)
VGG uses convlayers of size 3x3. Two remarks here: 1.) 3x3 is not a multiple of 8 which is a prerequisite of MP. 2.) Even 8x8 might not yield any performance gain, because it's simply too small. 

> **I've figured out the problem is related to bilinear upsampling (inside Upsample2D or tf.image.resize) and it breaks the mixed-precision performance.**
How did you figure this out? According to this website [ClickMe](https://developer.nvidia.com/blog/accelerating-medical-image-segmentation-tensor-cores-tensorflow-2/), upsampling shouldn't be a problem. 

> **It is too bad because the same network implemented with another framework is ~7-8x faster**
Namely PyTorch? MP can even slow down training if not configured properly. 

Is your data casted to float 16? Is your Batchsize a multiple of 8? Is the image shape a multiple of 8 for both dimensions (width/height)?

**Comment by Hazarapet** [comment]
@Arktius 
Almost everything in my network is multiple of 8. Batch Size, Decoder Filter Count, etc. 
**Also, if I'm not mistaken, they don't require everything to be a multiple of 8 (for example kernel size or dilation rate). 
Then why do I need mixed-precision, if it requires only certain numbers for my network?**
What about the output of my network? e.g I have 6 classes, how to force them to make 8?

**Comment by Arktius** [comment]
'Almost'? What is not?

What do you mean? Why do you need MP when it restricts the developer in many cases? Well, the benefit of training a model 2-8 x faster is enough motivation to fulfil all prerequisites. Training a model in 5 days instead of 10 or 14 days can save a lot of time and thus money.

What's your activation function at the output layer? Sigmoid for the case where always one class is present? Why not just add another 2 classes at the output? There must be always zero then. 

Could you provide all the details of your model and your data?

**Comment by zaccharieramzi** [comment]
I noticed the same issue, but I don't have a reproducible example to share yet, however I can say that my network uses some bilinear upsampling (as well as a lot of other stuff).
More specifically, my training time is 6x longer for the epochs where no nans are present. Nans start appearing after roughly 2000 steps (which takes about 4 hours).

@Hazarapet I don't know if you had seen this, but indeed bilinear upsampling is a cause of inefficiency for mixed precision in TensorFlow as reported in [this issue](https://github.com/tensorflow/tensorflow/issues/41934). They should have fixed the problem for [tf 2.4](https://github.com/tensorflow/tensorflow/commit/67d15573a776119d5a544ed266dc2514ae13c3b5).

However, I [am still wondering ](https://github.com/tensorflow/tensorflow/issues/41934#issuecomment-743068647) if the Nan issue has the same cause, and therefore the same fix.
